1.regression_modelling/scripts/2.r2_figure.py

```diff
@@ -0,0 +1,657 @@
+    y_train = training_df[model_name]
+    y_test = testing_df[model_name]
+    y_holdout = holdout_df[model_name]
+
+    # Predict on training, testing, and holdout data
+    y_train_pred = model.predict(X_train)
+    y_test_pred = model.predict(X_test)
+    y_holdout_pred = model.predict(X_holdout)
+
+    # Calculate R2 scores
+    r2_train = r2_score(y_train, y_train_pred)
+    r2_test = r2_score(y_test, y_test_pred)
+    r2_holdout = r2_score(y_holdout, y_holdout_pred)
+
+    # Append results to the list, including 'Model type'
+    results.append({'Feature': model_name, 'Stain': stain, 'Data split': 'Training', 'R2 score': r2_train, 'Model type': model_type})
+    results.append({'Feature': model_name, 'Stain': stain, 'Data split': 'Testing', 'R2 score': r2_test, 'Model type': model_type})
+    results.append({'Feature': model_name, 'Stain': stain, 'Data split': 'Holdout', 'R2 score': r2_holdout, 'Model type': model_type})
+
+# Convert the results into a dataframe
+results_df = pd.DataFrame(results)
+
+# Remove rows where 'Feature' starts with 'Nuclei_Location' (not an actual feature)
+results_df = results_df[~results_df['Feature'].str.startswith('Nuclei_Location')]
+
+print(results_df.shape)
+results_df.head()
+
+
+# ## Generate box plot comparing testing data across models and splits
+
+# ### Set function
+
+# In[7]:
+
+
+# Function to create the plot for each stain
+def plot_r2(results_df, palette_dict, datasplit):
+    # Set the figure size
+    plt.figure(figsize=(14, 8))
+
+    # Create the horizontal box plot (one per Model type and Stain combination)
+    sns.boxplot(
+        data=results_df,
+        x='R2 score',
+        y='model_stain',  # model_stain on y-axis for horizontal comparison
+        dodge=True,  # Adjust box plots for multiple hues
+        fliersize=0,  # Hide the default outliers as we add scatter later
+        width=0.5  # Adjust the width to increase separation
+    )
+
+    # Add scatter plot of individual points within each box, colored by 'Measurement'
+    sns.stripplot(
+        data=results_df,
+        x='R2 score',
+        y='model_stain',  # Same axis for horizontal separation
+        hue='Measurement',  # Color by 'Measurement'
+        palette=palette_dict,  # Customize colors with your palette
+        jitter=True,  # Jitter the points slightly for better visibility
+        dodge=True,
+        size=8,  # Size of the scatter points
+        edgecolor='black',
+        linewidth=1,
+        alpha=0.8  # Transparency of the points
+    )
+
+    # Customize the plot
+    plt.title(f'R2 Scores by Model Type and Stain: {datasplit} data')
+    plt.xlabel('R2 Score')
+    plt.ylabel('Model Stain')
+
+    # Move legend outside of the plot
+    plt.legend(title='Measurement', bbox_to_anchor=(1.05, 1), loc='upper left')
+
+    # Improve layout
+    plt.tight_layout()
+
+    # Save
+    plt.savefig(f"./{figures_dir}/shuffled_versus_final_r2.png", dpi=600)
+
+    # Show plot
+    plt.show()
```

Comment:
I would recommend adding type hints for this function and possible a small doc string? Also consider adding a arg of the save path to the function so that global args are not required

Reply:
Ahh I forgot to add that. I will add it in now thanks!